### ISSUE TITLE
lasio changes STEP: adding decimals

### ISSUE THREAD

**Issue by geolr** [kinverarity1/lasio#404]
LAS with a header saying: 
`STEP.m     0.1524 : STEP`
Becomes 
`STEP.m 0.1524000000000001 : STEP`

Only header values were touched in processing, like UWI and adding a prefix to curvenames.
Then las.write to a new file.

Apparently lasio does a re-calculate of depth-step for the first two lines in the data.
I'd assume this has it's origin in floating point arthimetics: Doing the same calculation `<depth-line-2> minus <depth-line-1>` in an interactive Python session give the same result: 0.1524000000000001

Should lasio round the values to 4 decimal places?

**Comment by kinverarity1** [comment]
Hi @geolr thanks for raising this issue! I think there are a few improvements to be made in this area.  Certainly the floating point arithmetic effect there is a bug, I feel like we should represent the depth step at the same precision as present in the depth values themselves.

Your issue has also prompted me to open #405, as I don't think that in your case lasio should be attempting to recalculate the value in the STEP header line at all, as you are only modifying the header.

**Comment by geolr** [comment]
Just a quick note, since I updated: 
lasio-0.28 still has the bug

**Comment by dcslagel** [comment]
Hi @geolr ,

I did this trying to make a test case of this error,  but it passes:

```
def test_step_unchanged_by_write():
    las = read(egfn("2.0/sample_2.0.las"))
    las.well['UWI'] = "123456789"
    las.write('test.las', version=2.0)
    assert las.well['STEP'].value == -0.125
```

As you mentioned above the write function is calculating the STEP from the first two data depths.  In the case of this test it is doing:
```
STEP = 1669.875 - 1670.000   # -0.125
las.well['STEP'].value = STEP
```

- Do you have an example las file that will produce the error?
- How are you adding a prefix to curvenames?  

Thanks,
DC

**Comment by dcslagel** [comment]
analysis notes:

1. running the documentations 'create las file from scratch instructions doesn't show the error:
https://lasio.readthedocs.io/en/latest/building.html
```
import lasio
import numpy as np

testfn = 'testfile.las'
dstart = 10
dstop = 12
dstep = 0.5

las = lasio.las.LASFile()

depths = np.arange(dstart, dstop, dstep)
las.add_curve('DEPTH', depths, unit='m')
las.write(testfn)
```

Outputs:
```
...
~Well ------------------------------------------------------
STRT.m    10.0 : START DEPTH
STOP.m    11.5 : STOP DEPTH
STEP.m     0.5 : STEP
...
```

2. Adding one more digit to the fractional-part of the step number and we get a result like @geolr's example:
```
import lasio
import numpy as np

testfn = 'testfile.las'
dstart = 10
dstop = 12
dstep = 0.51

las = lasio.las.lasfile()

depths = np.arange(dstart, dstop, dstep)
las.add_curve('depth', depths, unit='m')
las.write(testfn)
```
import lasio
import numpy as np

testfn = 'testfile.las'
dstart = 10
dstop = 12
dstep = 0.51  # NOTE 0.51

las = lasio.las.lasfile()

depths = np.arange(dstart, dstop, dstep)
las.add_curve('depth', depths, unit='m')
las.write(testfn)
```

Output:
```
~Well ------------------------------------------------------
...
STEP.m 0.5099999999999998 : STEP
...
```

**Comment by geolr** [comment]
Thanks @dcslagel that is the behaviour. 
One of my wells: 
> ~Well ------------------------------------------------------
STRT.m            205.283 : Start Depth
STOP.m          1740.1034 : Stop Depth
STEP.m 0.1524000000000001 : Step
NULL.              -999.0 : Null Value
COMP.                     : Company

**Comment by dcslagel** [comment]
@geolr, 

I've updated the test in pull-request #431 to use the STRT, STOP, and STEP from your well: 205.283, 1740.1034, 0.1524.

**Comment by kinverarity1** [comment]
@dcslagel can this be closed now that #432 has been merged and released as [v0.29](https://github.com/kinverarity1/lasio/releases/tag/v0.29)?

**Comment by dcslagel** [comment]
Yes.  The test uses the data that @geolr provided so it works now for the case he reported.
